fabric-particles-v1/src/main/java/net/fabricmc/fabric/api/particles/FabricParticleTypes.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.particles;
+
+import net.minecraft.particle.DefaultParticleType;
+import net.minecraft.particle.ParticleEffect;
+import net.minecraft.particle.ParticleType;
+import net.minecraft.particle.ParticleEffect.Factory;
+
+/**
+ * Methods for creating particle types, both simple and using an existing attribute factory.
+ */
+public final class FabricParticleTypes {
+
+    /**
+     * Creates a new, default particle type for the given id.
+     *
+     * @param id The particle id.
+     */
+    public static DefaultParticleType simple() {
+        return simple(false);
+    }
+
+    /**
+     * Creates a new, default particle type for the given id.
+     *
+     * @param id The particle id.
+     * @param alwaysSpawn True to always spawn the particle regardless of distance.
+     */
+    public static DefaultParticleType simple(boolean alwaysSpawn) {
+        return new Simple(alwaysSpawn);
+    }
+
+    /**
+     * Creates a new particle type with a custom factory for packet/data serialization.
+     *
+     * @param id The particle id.
+     * @param factory     A factory for serializing packet data and string command parameters into a particle effect.
+     */
+    public static <T extends ParticleEffect> ParticleType<T> complex(ParticleEffect.Factory<T> factory) {
+        return complex(false, factory);
+    }
+
+    /**
+     * Creates a new particle type with a custom factory for packet/data serialization.
+     *
+     * @param id The particle id.
+     * @param alwaysSpawn True to always spawn the particle regardless of distance.
+     * @param factory     A factory for serializing packet data and string command parameters into a particle effect.
+     */
+    public static <T extends ParticleEffect> ParticleType<T> complex(boolean alwaysSpawn, ParticleEffect.Factory<T> factory) {
+        return new Complex<>(alwaysSpawn, factory);
+    }
+
+    // Constructor is (gasp!) protected
+    public static class Simple extends DefaultParticleType {
```

Comment:
This and `Complex` seem completely unnecessary - why not just use anonymous classes here, since it's hidden behind utility methods anyway?

Reply:
I don't much like Anon classes because I always feel like they're going to have some kind of performance impact over staticaly-defined ones. Having the named classes here also means they're available for anyone who wishes to use them.

Reply:
Anonymous classes are strictly the same as named ones, they are generated at compile time.

Reply:
What about situations like this?

```
static Cow createCow(String sound) {
   return new Cow() {
       @Override
       public String moo() { return sound; }
   }
}
```
(and also lambda closurs)? I don't know whether that causes it to generate a new class at every invocation.

Reply:
It doesn't. It will generate a new *instance* at each invocation, as expected, with that instance having its own copy of `sound`.